<a href="https://colab.research.google.com/github/Pritija11/workshop-1/blob/main/wrkshp_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy as np
import pandas as pd
def cost_function(X, Y, W):

  m = len(Y)
  cost = np.sum((X.dot(W)-Y) **2)/(2*m)
  return cost



In [3]:
def gradient_descent(X, Y, W, alpha, iterations):
  cost_history = [0] * iterations
  m = len(Y)
  for iteration in range(iterations):
    Y_pred = X.dot(W)
    loss = Y_pred - Y
    dw = (X.T.dot(loss))/(m)
    W_update = W - alpha * dw
    cost = cost_function(X, Y, W_update)
    cost_history[iteration] = cost
  return W_update, cost_history


In [4]:
def rmse(Y, Y_pred):

  rmse = np.sqrt(np.mean((Y - Y_pred) ** 2))
  return rmse


In [5]:
def r2(Y, Y_pred):
  mean_y = np.mean(Y)
  ss_tot = np.sum((Y - mean_y) ** 2)
  ss_res = np.sum((Y - Y_pred) ** 2)
  r2 = 1 - (ss_res / ss_tot)
  return r2


In [6]:
from sklearn.model_selection import train_test_split

def main():
  data = pd.read_csv("/content/drive/MyDrive/dataset/student.csv")
  X = data[['Math', 'Reading']].values
  Y = data['Writing'].values
  X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

  W = np.zeros(X_train.shape[1])
  alpha = 0.001
  iterations = 1000

  W_optimal, cost_history = gradient_descent(X_train, Y_train, W, alpha, iterations)

  Y_pred = np.dot(X_test, W_optimal)

  model_rmse = rmse(Y_test, Y_pred)
  model_r2 = r2(Y_test, Y_pred)

  print("Final Weights:", W_optimal)
  print("Cost History (First 10 iterations):", cost_history[:10])
  print("RMSE on Test Set:", model_rmse)
  print("R-Squared on Test Set:", model_r2)

if __name__ == "__main__":
      main()



Final Weights: [4.7978325  5.02019875]
Cost History (First 10 iterations): [np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246)]
RMSE on Test Set: 615.8660062668006
R-Squared on Test Set: -1514.236975362205


1. Did your Model Overfitt, Underfitts, or performance is acceptable.

When alpha = 0.00001, The model is underfitting as the RMSE value is very high and R-square is negative. The cost history is slao constant which states Gradient descent did not converge properly.

2. Experiment with different value of learning rate, making it higher and lower, observe the result.

alpha = 0.0001:
Final Weights: [0.47978325 0.50201988]
Cost History (First 10 iterations): [np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211)]
RMSE on Test Set: 6.092665443799174
R-Squared on Test Set: 0.851706281452247

When alpha value is increased from 0.00001 to 0.0001, the model perfomance improved. The RMSE reduced from 63.36 to 6.09, and the R² score increased from −15.04 to 0.85, indicating that the model learned effectively. So, the model perfomance is acceptable neither underfitting nor overfitting.

aplha = 0.001:
Final Weights: [4.7978325  5.02019875]
Cost History (First 10 iterations): [np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246)]
RMSE on Test Set: 615.8660062668006
R-Squared on Test Set: -1514.236975362205

The model is unstable and performs poorly.



To - Do - 1:
1. Read and Observe the Dataset.
2. Print top(5) and bottom(5) of the dataset {Hint: pd.head and pd.tail}.
3. Print the Information of Datasets. {Hint: pd.info}.
4. Gather the Descriptive info about the Dataset. {Hint: pd.describe}
5. Split your data into Feature (X) and Label (Y).

In [8]:
data = pd.read_csv("/content/drive/MyDrive/dataset/student.csv")

In [9]:
print(data.head())

   Math  Reading  Writing
0    48       68       63
1    62       81       72
2    79       80       78
3    76       83       79
4    59       64       62


In [10]:
print(data.tail())

     Math  Reading  Writing
995    72       74       70
996    73       86       90
997    89       87       94
998    83       82       78
999    66       66       72


In [11]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   Math     1000 non-null   int64
 1   Reading  1000 non-null   int64
 2   Writing  1000 non-null   int64
dtypes: int64(3)
memory usage: 23.6 KB
None


In [12]:
print(data.describe())

              Math      Reading      Writing
count  1000.000000  1000.000000  1000.000000
mean     67.290000    69.872000    68.616000
std      15.085008    14.657027    15.241287
min      13.000000    19.000000    14.000000
25%      58.000000    60.750000    58.000000
50%      68.000000    70.000000    69.500000
75%      78.000000    81.000000    79.000000
max     100.000000   100.000000   100.000000


In [13]:
X = data[['Math', 'Reading']]
Y = data['Writing']

print(X.head())
print(Y.head())

   Math  Reading
0    48       68
1    62       81
2    79       80
3    76       83
4    59       64
0    63
1    72
2    78
3    79
4    62
Name: Writing, dtype: int64


• To - Do - 2:
1. To make the task easier - let’s assume there is no bias or intercept.
2. Create the following matrices:

Y = WTX

• To - Do - 3:
1. Split the dataset into training and test sets.
2. You can use an 80-20 or 70-30 split, with 80% (or 70%) of the data used for training and the rest
for testing.

In [16]:
X = data[['Math', 'Reading']].values.T

Y = data['Writing'].values

d = X.shape[0]
W = np.zeros(d)

Y_pred = np.dot(W.T, X)

print("Shape of X (features x samples):", X.shape)
print("Shape of W (weights):", W.shape)
print("Shape of Y_pred (predictions):", Y_pred.shape)

print("First 5 predicted values:", Y_pred[:5])

Shape of X (features x samples): (2, 1000)
Shape of W (weights): (2,)
Shape of Y_pred (predictions): (1000,)
First 5 predicted values: [0. 0. 0. 0. 0.]


• To - Do - 3:
1. Split the dataset into training and test sets.
2. You can use an 80-20 or 70-30 split, with 80% (or 70%) of the data used for training and the rest
for testing.

In [18]:
import pandas as pd
from sklearn.model_selection import train_test_split

X = data[['Math', 'Reading']].values
Y = data['Writing'].values

X_train, X_test, Y_train, Y_test = train_test_split(
    X, Y, test_size=0.2, random_state=42
)


print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of Y_train:", Y_train.shape)
print("Shape of Y_test:", Y_test.shape)

Shape of X_train: (800, 2)
Shape of X_test: (200, 2)
Shape of Y_train: (800,)
Shape of Y_test: (200,)
